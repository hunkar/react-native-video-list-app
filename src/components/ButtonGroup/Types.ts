import {IconNode} from 'react-native-elements/dist/icons/Icon';
import {StyleProp, ViewStyle} from 'react-native';

//TODO Other styles can be customized. Can be added in the future.
/**
 * title: text of button
 * icon: icon of button
 * onPress: press event of button
 */
export type ButtonProps = {
  title: string;
  icon?: IconNode;
  onPress?: (() => void) | undefined;
};

/**
 * containerStyle: Styles for group container
 * buttons: array of ButtonProps
 */
export type ButtonGroupProps = {
  containerStyle?: StyleProp<ViewStyle>;
  buttons: ButtonProps[];
};
