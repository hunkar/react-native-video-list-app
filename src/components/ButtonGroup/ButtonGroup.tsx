import React from 'react';
import {StyleSheet, View} from 'react-native';
import {Button} from 'react-native-elements';
import {ButtonGroupProps, ButtonProps} from './Types';

/**
 * This component provides group of same type buttons.
 * Developed for decrease workload for same types.
 *
 * @param props ButtonGroupProps
 * @returns ReactNode
 */
const ButtonGroup = (
  props: ButtonGroupProps = {
    containerStyle: {},
    buttons: [],
  },
) => {
  const {containerStyle, buttons} = props;

  return (
    <View
      style={{
        ...styles.containerStyle,
        ...(containerStyle as object),
      }}>
      {buttons.map(
        (
          buttonItem: ButtonProps = {
            title: 'Button',
            icon: {},
            onPress: () => {
              console.log('Set event for button.');
            },
          },
          buttonIndex,
        ) => (
          <Button
            testID={`button_${buttonIndex}`}
            key={`button_${buttonIndex}`}
            title={buttonItem.title}
            icon={buttonItem.icon}
            iconContainerStyle={styles.iconContainerStyle}
            titleStyle={styles.titleStyle}
            buttonStyle={styles.buttonStyle}
            containerStyle={styles.buttonContainerStyle}
            onPress={buttonItem.onPress}
          />
        ),
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  containerStyle: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  buttonContainerStyle: {
    flex: 1,
  },
  buttonStyle: {
    backgroundColor: '#d6d6d6',
    borderColor: '#4a4a4a',
    borderWidth: 1,
    borderRadius: 0,
  },
  titleStyle: {
    color: '#4a4a4a',
  },
  iconContainerStyle: {
    marginRight: 10,
  },
});

export default ButtonGroup;
