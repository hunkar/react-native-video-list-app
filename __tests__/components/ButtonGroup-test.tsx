// Note: test renderer must be required after react-native.
// import {render} from '@testing-library/react';
import renderer, {act} from 'react-test-renderer';
import React from 'react';
import {
  ButtonGroupProps,
  ButtonProps,
} from '../../src/components/ButtonGroup/Types';
import ButtonGroup from '../../src/components/ButtonGroup/ButtonGroup';

test('snapshot equility', async () => {
  const firtButton: ButtonProps = {
    title: 'Select',
    onPress: jest.fn(),
  };
  const secondButton: ButtonProps = {
    title: 'Cancel',
    onPress: jest.fn(),
  };

  const mockProps: ButtonGroupProps = {
    buttons: [firtButton, secondButton],
  };

  const component = renderer.create(<ButtonGroup {...mockProps} />);

  expect(component.toJSON()).toMatchSnapshot();
});

test('renders title, onPress correctly', async () => {
  const firtButton: ButtonProps = {
    title: 'Select',
    onPress: jest.fn(),
  };
  const secondButton: ButtonProps = {
    title: 'Cancel',
    onPress: jest.fn(),
  };

  const mockProps: ButtonGroupProps = {
    buttons: [firtButton, secondButton],
  };

  const component = renderer.create(<ButtonGroup {...mockProps} />);

  const firstButtonComponent = component.root.findByProps({
    testID: 'button_0',
  }).props;
  const secondButtonComponent = component.root.findByProps({
    testID: 'button_1',
  }).props;

  expect(firstButtonComponent.title).toEqual(mockProps.buttons[0].title);
  expect(firstButtonComponent.onPress).toEqual(mockProps.buttons[0].onPress);

  expect(secondButtonComponent.title).toEqual(mockProps.buttons[1].title);
  expect(secondButtonComponent.onPress).toEqual(mockProps.buttons[1].onPress);
});

test('renders title, onPress correctly', async () => {
  const firtButton: ButtonProps = {
    title: 'Select',
    onPress: jest.fn(),
  };
  const secondButton: ButtonProps = {
    title: 'Cancel',
    onPress: jest.fn(),
  };

  const mockProps: ButtonGroupProps = {
    buttons: [firtButton, secondButton],
  };

  const component = renderer.create(<ButtonGroup {...mockProps} />);

  const firstButtonComponent = component.root.findByProps({
    testID: 'button_0',
  }).props;
  const secondButtonComponent = component.root.findByProps({
    testID: 'button_1',
  }).props;

  expect(firstButtonComponent.title).toEqual(mockProps.buttons[0].title);
  expect(firstButtonComponent.onPress).toEqual(mockProps.buttons[0].onPress);

  expect(secondButtonComponent.title).toEqual(mockProps.buttons[1].title);
  expect(secondButtonComponent.onPress).toEqual(mockProps.buttons[1].onPress);
});

test('call onPress functions correct', async () => {
  const firtButton: ButtonProps = {
    title: 'Select',
    onPress: jest.fn(),
  };
  const secondButton: ButtonProps = {
    title: 'Cancel',
    onPress: jest.fn(),
  };

  const mockProps: ButtonGroupProps = {
    buttons: [firtButton, secondButton],
  };

  const component = renderer.create(<ButtonGroup {...mockProps} />);

  const firstButtonComponent = component.root.findByProps({
    testID: 'button_0',
  }).props;
  const secondButtonComponent = component.root.findByProps({
    testID: 'button_1',
  }).props;

  act(() => {
    firstButtonComponent.onPress();
    secondButtonComponent.onPress();
  });

  expect(firtButton.onPress).toBeCalledTimes(1);
  expect(secondButton.onPress).toBeCalledTimes(1);
});