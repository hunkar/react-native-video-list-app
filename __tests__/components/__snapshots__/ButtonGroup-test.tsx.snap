// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot equility 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flexDirection": "row",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        Object {
          "flex": 1,
        },
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="button_0"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#d6d6d6",
            "borderColor": "#4a4a4a",
            "borderRadius": 0,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4a4a4a",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Select
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        Object {
          "flex": 1,
        },
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="button_1"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#d6d6d6",
            "borderColor": "#4a4a4a",
            "borderRadius": 0,
            "borderWidth": 1,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#4a4a4a",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Cancel
        </Text>
      </View>
    </View>
  </View>
</View>
`;
